File: solutions/elixir/039-n-queens.ex
# Problem 39: N-Queens (LeetCode 51)
# Difficulty: Hard
# Language: Elixir
# 
defmodule NQueens do
  def solve_n_queens(n) do
    go(0, n, MapSet.new(), MapSet.new(), MapSet.new(), [])
    |> Enum.map(fn board ->
      Enum.map(board, fn c ->
        String.duplicate(".", c) <> "Q" <> String.duplicate(".", n - c - 1)
      end)
    end)
  end

  defp go(n, n, _cols, _d1, _d2, board), do: [Enum.reverse(board)]
  defp go(row, n, cols, d1, d2, board) do
    Enum.flat_map(0..(n - 1), fn c ->
      if not MapSet.member?(cols, c) and
         not MapSet.member?(d1, row - c) and
         not MapSet.member?(d2, row + c) do
        go(row + 1, n, MapSet.put(cols, c),
           MapSet.put(d1, row - c), MapSet.put(d2, row + c), [c | board])
      else
        []
      end
    end)
  end
end

File: solutions/scala/039-n-queens.scala
/**
 * Problem 39: N-Queens (LeetCode 51)
 * Difficulty: Hard
 * Language: Scala
 */
def solveNQueens(n: Int): List[List[String]] = {
  def go(row: Int, cols: Set[Int], d1: Set[Int], d2: Set[Int],
         board: List[Int]): List[List[String]] = {
    if (row == n) List(board.reverse.map(c =>
      "." * c + "Q" + "." * (n - c - 1)))
    else (0 until n).toList
      .filter(c => !cols(c) && !d1(row - c) && !d2(row + c))
      .flatMap(c =>
        go(row + 1, cols + c, d1 + (row - c), d2 + (row + c), c :: board))
  }
  go(0, Set(), Set(), Set(), Nil)
}
